In [19]:
import numpy as np
import random

In [20]:
data = np.random.randint(0, 50, [4,3,2,3])
print(data)
print(data.sum(axis=0))
print(np.ndarray((3,4)))

[[[[ 7  7 48]
   [28 15 12]]

  [[24 17 29]
   [21  1 32]]

  [[ 2  2 16]
   [48 36 28]]]


 [[[39 28 29]
   [46  5 33]]

  [[31  4 27]
   [10 13 17]]

  [[34  2 11]
   [45  0 26]]]


 [[[35 27 17]
   [33  6 30]]

  [[10 22 40]
   [23 43  4]]

  [[35 22 27]
   [20 39 25]]]


 [[[44 30  5]
   [13 48  7]]

  [[29  3 26]
   [11 25 32]]

  [[44 34 20]
   [19 11 46]]]]
[[[125  92  99]
  [120  74  82]]

 [[ 94  46 122]
  [ 65  82  85]]

 [[115  60  74]
  [132  86 125]]]
[[6.93831833e-310 6.93831833e-310 4.94065646e-324 4.94065646e-324]
 [0.00000000e+000 0.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 1.63041663e-322 4.65412647e-310 4.65412633e-310]]


In [21]:
np.ndim(data)

4

In [22]:
arr = np.random.randint(1, 10, size = (3, 4))
print(arr)

[[2 7 4 3]
 [6 3 6 8]
 [1 8 4 4]]


In [23]:
res_sum = arr.sum(axis = 0)  # 聚合操作
print(res_sum)

[ 9 18 14 15]


In [24]:
res_sum = arr.sum(axis = 1)  # 聚合操作
print(res_sum)

[16 23 17]


In [25]:
arr.min(axis = 0)

array([1, 3, 4, 3])

In [26]:
arr.prod(axis = 0) # 乘积 axis = None 所有的乘积 笛卡尔乘积

array([ 12, 168,  96,  96])

In [27]:
arr.mean(axis = 1)  # 求平均

array([4.  , 5.75, 4.25])

In [28]:
arr.std(axis = 1) # 标准偏差  https://www.cnblogs.com/lonelyshy/p/12531544.html

array([1.87082869, 1.78535711, 2.48746859])

In [29]:
arr.std(axis=0)

array([2.1602469 , 2.1602469 , 0.94280904, 2.1602469 ])

In [30]:
arr.argmin(axis = 1) # 加了arg 表示返回最小值的索引

array([0, 1, 0])

# 矩阵运算

In [31]:
arr + 1  # 每个元素加1

array([[3, 8, 5, 4],
       [7, 4, 7, 9],
       [2, 9, 5, 5]])

In [32]:
arr2 = np.random.randint(1, 10, size = (4, 3))
arr2

array([[6, 9, 7],
       [1, 1, 2],
       [4, 5, 1],
       [5, 9, 2]])

In [33]:
print(arr)
# arr + arr2 


[[2 7 4 3]
 [6 3 6 8]
 [1 8 4 4]]


In [34]:
#arr * arr2  # np.dot  普通乘法

In [35]:
# 要求第一个矩阵的列数和第二个矩阵的行数相同才能做矩阵点乘
# 矩阵的积是有顺序的，不满足乘法交换律
np.dot(arr, arr2) # 矩阵积 点乘  

array([[ 50,  72,  38],
       [103, 159,  70],
       [ 50,  73,  35]])

In [36]:
arr.argmin()

8

In [37]:
def select_sort(select_array):
    for i in  range(0, len(select_array) - 1):
        index = np.argmin(select_array[i:len(select_array)]) + i # slices 后是一个新数组和原来的不是同地址
        print(index)
        if select_array[i] > select_array[index]:
            select_array[i], select_array[index] = select_array[index], select_array[i]
select_array = [3, 6, 2, 7, 5, 9, 1]
select_sort(select_array)
print(select_array)

6
2
6
4
6
6
[1, 2, 3, 5, 6, 7, 9]


In [38]:
arr = np.arange(0, 2, 0.3)

In [39]:
type(arr)

numpy.ndarray

In [40]:
arr

array([0. , 0.3, 0.6, 0.9, 1.2, 1.5, 1.8])

In [41]:
arr = np.linspace(0, 10)

In [42]:
arr

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

In [43]:
arr = np.arange(12).reshape(3,4)
arr 

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [44]:
a = np.arange(12)
a

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [45]:
b = a 
b

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [46]:
b.shape = 3,4

In [47]:
b

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [48]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [49]:
a.shape

(3, 4)

In [50]:
c = a.view()

In [51]:
c

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [52]:
c.shape = 2, 6 
c

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11]])

In [53]:
a.shape 

(3, 4)

In [54]:
c[0,4] = 1234

In [55]:
a

array([[   0,    1,    2,    3],
       [1234,    5,    6,    7],
       [   8,    9,   10,   11]])

In [56]:
c

array([[   0,    1,    2,    3, 1234,    5],
       [   6,    7,    8,    9,   10,   11]])

In [57]:
s = a[ : , 1:3]  

In [58]:
s[:] = 10 

In [59]:
s

array([[10, 10],
       [10, 10],
       [10, 10]])

In [60]:
a

array([[   0,   10,   10,    3],
       [1234,   10,   10,    7],
       [   8,   10,   10,   11]])

In [61]:
a_test = np.arange(12)

In [62]:
a_test 

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [63]:
cc = a_test.reshape(3,4)

In [64]:
cc

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [65]:
cc[0, 3] = 1234
#ss = cc[:, 1:3]

In [66]:
cc

array([[   0,    1,    2, 1234],
       [   4,    5,    6,    7],
       [   8,    9,   10,   11]])

In [67]:
ss = cc[:, 1:3]
ss

array([[ 1,  2],
       [ 5,  6],
       [ 9, 10]])

In [68]:
ss[:] = 10
ss

array([[10, 10],
       [10, 10],
       [10, 10]])

In [69]:
cc

array([[   0,   10,   10, 1234],
       [   4,   10,   10,    7],
       [   8,   10,   10,   11]])

In [70]:
br = np.ones((2,3))
br

array([[1., 1., 1.],
       [1., 1., 1.]])

In [71]:
br2 = np.arange(3)

In [72]:
br2 

array([0, 1, 2])

In [73]:
br + br2

array([[1., 2., 3.],
       [1., 2., 3.]])

In [74]:
br*10

array([[10., 10., 10.],
       [10., 10., 10.]])

In [75]:
# https://www.cnblogs.com/Shawnyi/p/10370815.html

numpy.dot(a, b, out=None)传入2个数组，然后返回两个数组的点积。

根据官方文档https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html

可以分为以下情况：

1. 如果处理的是一维数组，则得到的是两数组的內积

2. 如果是二维数组（矩阵）之间的运算，则得到的是矩阵积

3. 如果是0维数组(标量)，则得到的是两个数的乘积

4. 如果a是N维数组，b是一维数组，a中最后一个轴(如(x, y, z)中的z， 即列)，每一列分别与b求内积

5. 如果a是N维数组，b是M(M>=2)维数组，由a中的最后一个轴(即列)的元素与b中从第二到最后轴中对应元素求乘积之和


numpy.dot
numpy.dot(a， b，出 = 无)
两个数组的点积。特别

如果a和b都是一维数组，则它是矢量的内部乘积（没有复杂的偶联）。

如果a和b都是二维数组，则它是矩阵乘法，但使用matmul或是首选。a @ b

如果a或b是 0-D（标量），则等效于乘法和使用或首选。numpy.multiply(a, b)a * b

如果a是 N-D 数组，b是一个一维数组，则它是a和b的最后一个轴上的求和乘积。

如果a是 N-D 数组，b是 M-D 数组（其中），则它是b的最后一个轴和倒数第二个轴上的求和乘积：M>=2

dot(a, b)[i,j,k,m] = sum(a[i,j,:] * b[k,:,m])
参数
array_like
第一个参数。

barray_like
第二个参数。

出ndarray，可选
输出参数。它必须具有不使用时将返回的确切类型。特别是，它必须具有正确的类型，必须是 C 连续的，并且其 dtype 必须是要返回的dot（a，b） dtype。这是一个性能功能。因此，如果不满足这些条件，将引发异常，而不是尝试灵活。

返回
输出ndarray
返回a和b的点积。如果a和b都是标量或两个一维数组，则返回标量;否则将返回数组。如果给出出，则返回它。

提高
值错误
如果a的最后一个维度的大小与b的倒数第二个维度的大小不同。

In [76]:
arr6 = np.arange(1,13).reshape(3,2,2)
arr7 = np.arange(12, 24).reshape(2,2,3)
print(arr6)
print("\n\n")
print(arr7)
np.dot(arr6, arr7)

[[[ 1  2]
  [ 3  4]]

 [[ 5  6]
  [ 7  8]]

 [[ 9 10]
  [11 12]]]



[[[12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]]]


array([[[[ 42,  45,  48],
         [ 60,  63,  66]],

        [[ 96, 103, 110],
         [138, 145, 152]]],


       [[[150, 161, 172],
         [216, 227, 238]],

        [[204, 219, 234],
         [294, 309, 324]]],


       [[[258, 277, 296],
         [372, 391, 410]],

        [[312, 335, 358],
         [450, 473, 496]]]])

计算过程如下：

42 = 1 * 12 + 2 * 15

45 = 1 * 13 + 2 * 16

48 = 1 * 14 + 2 * 17

...

159 = 10 * 6 + 11 * 9

180 = 10 * 7 + 11 * 10

201 = 10 * 8 + 11 * 11

split(ary, indices_or_sections, axis=0)
函数功能：
把数组从左到右按顺序切分
参数说明：
ary，要切分的数组
indices_or_sections，如果该参数是一个整数n，就是把数组ary平均切分成n份；如果该参数是一个数组a，则以参数数组a中的元素为轴切分数组ary（左开右闭）
axis，表示沿哪个维度切分数组。默认值为0，代表横向切分ary，相当于hsplit；当其为1时，代表纵向切分ary，相当于vsplit。

特殊用法举例：
x, y = np.split(data, (4,), axis=1)
把数据集data中的所有数据在第四、五列之间分割，这在机器学习中分离标签是经常使用。

另外，使用split平均切分数组时，如果不能均分会报如下错误：
ValueError: array split does not result in an equal division
此时可用 array_split代替 split

In [77]:
split_array = np.arange(1, 10).reshape(3, 3)
print(split_array)
split_res = np.split(split_array,[1, 3], axis = 1)
print(split_res)
print(type(split_res))
for i in split_res:
    print(i)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[array([[1],
       [4],
       [7]]), array([[2, 3],
       [5, 6],
       [8, 9]]), array([], shape=(3, 0), dtype=int64)]
<class 'list'>
[[1]
 [4]
 [7]]
[[2 3]
 [5 6]
 [8 9]]
[]


In [78]:
arr8 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8])
np.split(arr8, [2, 7], axis = 0)

[array([0, 1]), array([2, 3, 4, 5, 6]), array([7, 8])]

In [79]:
np_split = np.random.randint(1, 21, size = (4, 5))
print(np_split)
new_np_split = np.split(np_split, [1, 3], axis = 0)
print(type(new_np_split))
print(new_np_split)

[[15 17 15 11  9]
 [18 17  3  4  6]
 [ 5 17 16  3  4]
 [ 1 12 11  5  9]]
<class 'list'>
[array([[15, 17, 15, 11,  9]]), array([[18, 17,  3,  4,  6],
       [ 5, 17, 16,  3,  4]]), array([[ 1, 12, 11,  5,  9]])]


# python broadcasting
    三、Broadcast（广播）的规则
    All input arrays with ndim smaller than the input array of largest ndim, have 1’s prepended to their shapes.
    The size in each dimension of the output shape is the maximum of all the input sizes in that dimension.
    An input can be used in the calculation if its size in a particular dimension either matches the output size in that dimension, or has value exactly 1.
    If an input has a dimension size of 1 in its shape, the first data entry in that dimension will be used for all calculations along that dimension. In other words, the stepping machinery of the ufunc will simply not step along that dimension (the stride will be 0 for that dimension).
    翻译如下：

    让所有输入数组都向其中shape最长的数组看齐，shape中不足的部分都通过在前面加1补齐
    输出数组的shape是输入数组shape的各个轴上的最大值
    如果输入数组的某个轴和输出数组的对应轴的长度相同或者其长度为1时，这个数组能够用来计算，否则出错
    当输入数组的某个轴的长度为1时，沿着此轴运算时都用此轴上的第一组值
    来看更为一般的broadcasting rules：
    当操作两个array时，numpy会逐个比较它们的shape（构成的元组tuple），只有在下述情况下，两arrays才算兼容：

    相等
    其中一个为1，（进而可进行拷贝拓展已至，shape匹配）
    下面通过实际例子来解释说明上述的四条规则：（下面例子均来自于numpy 中的 broadcasting（广播）机制）
    https://zhuanlan.zhihu.com/p/35010592
    https://www.jb51.net/article/182600.htm  *********
    https://www.jianshu.com/p/fbf79b761005  **********
    吴恩达深度学习笔记(14)-Python中的广播（Broadcasting in Python）
    用书中的话来介绍广播的规则：两个数组之间广播的规则：如果两个数组的后缘维度（即从末尾开始算起的维度）的轴长度相等或其中一方的长度为1，则认为他们是广播兼容的，广播会在缺失和(或)长度为1的维度上进行。************************

In [80]:
broad_array = np.random.randint(1, 21, size = (4, 5))
broad_array_one = np.random.randint(1, 5, size = (5,))
print(broad_array)
print('--' * 30)
print(broad_array_one)
print('--' * 30)
print(broad_array + broad_array_one)
print('--' * 30)
print(broad_array * broad_array_one)

[[19 10  5 13  7]
 [ 1 16  5 11 16]
 [20  4 18  2 15]
 [ 8  4 19 19  7]]
------------------------------------------------------------
[2 4 1 3 4]
------------------------------------------------------------
[[21 14  6 16 11]
 [ 3 20  6 14 20]
 [22  8 19  5 19]
 [10  8 20 22 11]]
------------------------------------------------------------
[[38 40  5 39 28]
 [ 2 64  5 33 64]
 [40 16 18  6 60]
 [16 16 19 57 28]]


In [81]:
arr = np.arange(12).reshape(4,3)
print(arr)
print(arr.mean(0))
arr-arr.mean(0)   # 4.5  5.5  6.5 broadcasting 补3个

[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
[4.5 5.5 6.5]


array([[-4.5, -4.5, -4.5],
       [-1.5, -1.5, -1.5],
       [ 1.5,  1.5,  1.5],
       [ 4.5,  4.5,  4.5]])